lib/src/main/java/macrobase/datamodel/DataFrame.java

```diff
@@ -0,0 +1,311 @@
+    public DataFrame() {
+        this.schema = new Schema();
+        this.stringCols = new ArrayList<>();
+        this.doubleCols = new ArrayList<>();
+        this.indexToTypeIndex = new ArrayList<>();
+        this.numRows = 0;
+    }
+
+    /**
+     * @return shallow copy of dataframe
+     */
+    public DataFrame copy() {
+        DataFrame other = new DataFrame();
+        other.schema = schema.copy();
+        other.stringCols = new ArrayList<>(stringCols);
+        other.doubleCols = new ArrayList<>(doubleCols);
+        other.indexToTypeIndex = new ArrayList<>(indexToTypeIndex);
+        other.numRows = numRows;
+        return other;
+    }
+
+    public Schema getSchema() {return this.schema;}
+    public int getNumRows() {return numRows;}
+    public ArrayList<double[]> getDoubleCols() { return doubleCols; }
+    public ArrayList<String[]> getStringCols() { return stringCols; }
+
+    // Fast Column-based methods
+    private void addDoubleColumnInternal(double[] colValues) {
+        doubleCols.add(colValues);
+        indexToTypeIndex.add(doubleCols.size()-1);
+    }
+    public DataFrame addDoubleColumn(String colName, double[] colValues) {
+        if (numRows == 0) {
+            numRows = colValues.length;
+        }
+        schema.addColumn(ColType.DOUBLE, colName);
+        addDoubleColumnInternal(colValues);
+        return this;
+    }
+    private void addStringColumnInternal(String[] colValues) {
+        stringCols.add(colValues);
+        indexToTypeIndex.add(stringCols.size()-1);
+    }
+    public DataFrame addStringColumn(String colName, String[] colValues) {
+        if (numRows == 0) {
+            numRows = colValues.length;
+        }
+        schema.addColumn(ColType.STRING, colName);
+        addStringColumnInternal(colValues);
+        return this;
+    }
+
+    protected int[] getSubIndices(List<Integer> columns) {
+        int d = columns.size();
+        int[] typeSubIndices = new int[d];
+        for (int i = 0; i < d; i++) {
+            typeSubIndices[i] = indexToTypeIndex.get(columns.get(i));
+        }
+        return typeSubIndices;
+    }
+
+    public double[] getDoubleColumn(int columnIdx) {
+        return doubleCols.get(indexToTypeIndex.get(columnIdx));
+    }
+    public double[] getDoubleColumnByName(String columnName) {
+        return doubleCols.get(indexToTypeIndex.get(schema.getColumnIndex(columnName)));
+    }
+    public ArrayList<double[]> getDoubleCols(List<Integer> columns) {
+        ArrayList<double[]> cols = new ArrayList<>();
+        for (int c : columns) {
+            cols.add(getDoubleColumn(c));
+        }
+        return cols;
+    }
+    public ArrayList<double[]> getDoubleColsByName(List<String> columns) {
+        return getDoubleCols(this.schema.getColumnIndices(columns));
+    }
+    public String[] getStringColumn(int columnIdx) {
+        return stringCols.get(indexToTypeIndex.get(columnIdx));
+    }
+    public String[] getStringColumnByName(String columnName) {
```

Comment:
Want to keep the API very lightweight, people aren't buying into the MB ecosystem they just want to plug and play.